### PR TITLE
[ Dataset ] Add Directory Dataset Producer @open sesame 03/24 20:22

### PR DESCRIPTION
**Pull request by jijoongmoon**
This pr includes directory dataset producer for bmp image only.

We can extend the type of image to jpg, png with the proper decoder and add a resizer according to requirements.

C API and unit test cases will be added later PRs.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: nntrainer/dataset/dir_data_producers.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2023 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file   dir_data_producers.h
+ * @date   24 Feb 2023
+ * @brief  This file contains dir data producers, reading from the fiels in
```
Comment (DonghakPark) [inline nntrainer/dataset/dir_data_producers.h:7]: Typo : ```fiels``` --> ```files```

**Review by DonghakPark — APPROVED** [review]
LGTM!

File: nntrainer/dataset/dir_data_producers.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2023 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file   dir_data_producers.h
+ * @date   24 Feb 2023
+ * @brief  This file contains dir data producers, reading from the fiels in
```
Comment (DonghakPark) [inline nntrainer/dataset/dir_data_producers.cpp:7]: Typo : fiels --> files

**Review by DonghakPark — APPROVED** [review]
LGTM!

File: nntrainer/dataset/dir_data_producers.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/dataset/dir_data_producers.cpp]: fixed. Thanks.

File: nntrainer/dataset/dir_data_producers.h (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/dataset/dir_data_producers.h]: fixed. Thanks.

File: nntrainer/dataset/dir_data_producers.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    NNTR_THROW_IF(result != row_padded, std::invalid_argument)
+      << "Cannot read bmp pixel data";
+
+    for (uint j = 0; j < width; j++) {
+
+      input[height * i + j] = (float)data[j * 3 + 2];
+
+      input[(height * width) + height * i + j] = (float)data[j * 3 + 1];
+
+      input[(height * width) * 2 + height * i + j] = (float)data[j * 3];
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/dataset/dir_data_producers.cpp:69]: It supposes to be BGR --> RGB.

File: nntrainer/dataset/dir_data_producers.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  size_t result = fread(info, sizeof(unsigned char), 54, f);
+  NNTR_THROW_IF(result != 54, std::invalid_argument)
+    << "Cannot read bmp header";
+
+  uint w = *(int *)&info[18];
+  uint h = *(int *)&info[22];
+
+  size_t row_padded = (width * 3 + 3) & (~3);
+  unsigned char *data = new unsigned char[row_padded];
+  unsigned char tmp;
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/dataset/dir_data_producers.cpp]: fixed. thanks.

File: nntrainer/dataset/dir_data_producers.cpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+ * remain as TODO
+ * @param     path file path
+ * @param     inputs float data for image pixel
+ * @param     width width
+ * @param     height height
+ */
+static void readImage(const std::string path, float *input, uint width,
+                      uint height) {
+  std::ifstream ifs;
+  FILE *f = fopen(path.c_str(), "rb");
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/dataset/dir_data_producers.cpp:40]: fixed. thanks.

**Pull request merged**